File: 03405. Count the Number of Arrays with K Matching Adjacent Elements/solution.js
/**
 * Problem: 3405. Count the Number of Arrays with K Matching Adjacent Elements
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

// Modulo constant for calculations
const MOD = 1000000007n

/**
 * Computes the modular exponentiation using exponentiation by squaring
 *
 * @param {bigint} base - The base number
 * @param {number} exponent - The exponent to raise the base to
 *
 * @returns {bigint} - The result of (base ^ exponent) % MOD
 */
const power = (base, exponent) =>
  exponent & 1
    ? (power(base, exponent - 1) * base) % MOD
    : exponent
    ? power((base * base) % MOD, exponent >> 1)
    : 1n

/**
 * Computes the binomial coefficient C(n, k) modulo MOD
 *
 * @param {number} n - Total number of items
 * @param {number} k - Number of items to choose
 *
 * @returns {bigint} - The binomial coefficient C(n, k) % MOD
 */
const combination = (n, k) =>
  k > n
    ? 0n
    : (((factorials[n] * inverseFactorials[k]) % MOD) *
        inverseFactorials[n - k]) %
      MOD

// Initialize factorial calculation
let factorial = 1n
// Precompute factorials up to 100000
const factorials = [...Array(1e5)].map((x, i) =>
  i ? (factorial = (factorial * BigInt(i)) % MOD) : factorial
)

// Set exponent for modular inverse calculation
factorial = Number(MOD) - 2

// Precompute modular inverses of factorials
const inverseFactorials = factorials.map((x) => power(x, factorial))

/**
 * Counts arrays of length n with m possible values and exactly k adjacent equal pairs
 *
 * @param {number} n - Array length
 * @param {number} m - Number of possible values per element
 * @param {number} k - Required number of adjacent equal pairs
 *
 * @returns {number} - Count modulo 1000000007
 */
const countGoodArrays = (n, m, k) =>
  Number(
    (((combination(n - 1, k) * BigInt(m)) % MOD) *
      power(BigInt(m - 1), n - k - 1)) %
      MOD
  )

File: 03405. Count the Number of Arrays with K Matching Adjacent Elements/solution.go
/**
 * Problem: 3405. Count the Number of Arrays with K Matching Adjacent Elements
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

// Modulo constant for calculations
const (
	MOD = 1e9 + 7
	MX  = 100000
)

var (
	// Precomputed factorials for combination calculations
	factorials    = make([]int64, MX)
	// Precomputed inverse factorials for combination calculations
	inverseFactorials = make([]int64, MX)
)

// Fast exponentiation with modular arithmetic
func qpow(x int64, n int) int64 {
	// Compute x^n mod MOD using binary exponentiation for efficient modular arithmetic
	res := int64(1)

	for n > 0 {
		if n&1 == 1 {
			res = res * x % MOD
		}

		x = x * x % MOD
		n >>= 1
	}

	return res
}

// Initialize factorials and inverse factorials arrays
func initFact() {
	// Check if factorials are already computed to avoid recomputation
	if factorials[0] != 0 {
		return
	}

	// Initialize factorial[0] = 1
	factorials[0] = 1

	// Compute factorials: factorial[i] = i!
	for i := 1; i < MX; i++ {
		factorials[i] = factorials[i-1] * int64(i) % MOD
	}

	// Compute inverse of the largest factorial using Fermat's little theorem
	inverseFactorials[MX-1] = qpow(factorials[MX-1], MOD-2)

	// Compute remaining inverse factorials using the relation: inv[i-1] = inv[i] * i
	for i := MX - 1; i > 0; i-- {
		inverseFactorials[i-1] = inverseFactorials[i] * int64(i) % MOD
	}
}

// Calculate combination C(n,m) using precomputed factorials
func comb(n, m int) int64 {
	return factorials[n] * inverseFactorials[m] % MOD * inverseFactorials[n-m] % MOD
}

// Count arrays with exactly k matching adjacent elements
func countGoodArrays(n, m, k int) int {
	initFact()

	return int(comb(n-1, k) * int64(m) % MOD * qpow(int64(m-1), n-k-1) % MOD)
}
